File: src/main/java/io/github/justfoxx/teacup/v1/event/Event.java
package io.github.justfoxx.teacup.v1.event;

/**
 * An event that can be invoked
 * @param <V> The type of the value
 * @param <I> The type of the invoker additional data
 */
public interface Event <V,I> {
    void invoke(I value);

    /**
     * Use it only when generic type I is Void
     */
    default void invoke() {
        invoke(null);
    }
}

File: src/main/kotlin/io/github/justfoxx/teacup/v1/event/Event.kt
package io.github.justfoxx.teacup.v1.event

/**
 * An event that can be invoked.
 * @param Value The type of the value passed to the event.
 * @param Invoker The type of the invoker lambda.
 */
interface Event<Value, Invoker> {
    fun invoker(): Invoker
    fun getAll(): Collection<Value>
}
